### ISSUE TITLE
Nested-TEA pattern

### ISSUE THREAD

**Issue by eratio08** [Janiczek/elm-store-pattern#2]
How would this be applied to a project that uses the nested TEA pattern? I fiddled a bit with it but I'm not too happy with the result. Do you know of any example you could link to?

**Comment by Janiczek** [comment]
Hey @eratio08, can you summarize a bit what do you mean by a nested TEA pattern?

If you mean nested Model/View/Update components, then you can still just pass the Store around as argument to various functions (update, view, subscriptions most likely), nothing changes about that.

In case you mean something like separate Elm applications working together, then I have no experience of moving data between them. Perhaps via webcomponent attributes...

**Comment by eratio08** [comment]
Hey @Janiczek,

I was referring to [Nested TEA pattern](http://sporto.github.io/elm-patterns/architecture/nested-tea.html). So I tried this with [a root model like shows in the Elm SPA Example by Richard](https://github.com/rtfeldman/elm-spa-example/blob/cb32acd73c3d346d0064e7923049867d8ce67193/src/Main.elm#L37-L46) and I tried to centralize data sending and store change interception to the main TEA. If I see it correctly one has to use the [translator pattern](http://sporto.github.io/elm-patterns/architecture/translator.html) to allow for communication of the nested TEA to the root TEA. This feels like breaking isolation/ leaking abstraction to the root TEA. That the main reason to ask if there is any example of this to see if this is just how it will turn out or if there is some other neat trick around this.

Thanks for your answer!

**Comment by erlandsona** [comment]
@eratio08 class is in session 😅 

For me, these patterns have started falling into a spectrum that I think a lot of people are referring to or talking about but aren't detailing the transitions between.

One thing the sporto site misses out on is when/where to leverage any of these patterns.

For us it seems that you should always start with the simplest thing that can work so signatures that only depend on state ala:
`toggle : { checkbox : Bool } -> Html msg`
**as this gets more complex** it's fairly straightforward to "upgrade it with a translator when you want this to "communicate it's events to the parent that's rendering it" to something like
`toggle : { checkbox : Bool, onClick : Bool -> msg } -> Html msg` <- notice we're still in whatever `msg` Elm infers for the "parent" in this case.
**as this gets more complex** that record will likely become it's own bespoke `Args` or `Config` type at which point you'll likely be applying the "Builder" pattern to that object to opt-in to certain behaviors while keeping a 'simpler' exposed API for call sites ala: 
```elm
type alias Config msg = { bg : Color, onClick : Maybe msg } -- we have it if onClick is Nothing then the button is disabled & greyed out.

defaultConf = { bg = Default, onClick = Nothing }

onPress msg conf = { conf | onClick = Just msg } -- just showing Builder here as an example.

button : Config msg -> Html msg -- notice we're still forcing the parent to define & handle their own Msg constructors for these elements.
```
**as it gets even more complex** The builder starts being responsible for "too many arguments" and the callers are having to define their own handlers for each of the different events that are being "exposed" via the `Config` via `Something -> msg` keys. From here we apply a thing called
# Defunctionalization 🙌 

Which ironically we end up with just a regular `Cmd.map ChildMsg` sorta situation again except this time we know what our exposed api of events _should_ actually be because we went through the growing pains of the Builder pattern. 

## How it works

Take every `Something -> msg` function in the `Builder` record (and arguably you don't need this unless you already have more than one or two "events" you need the caller to Handle.) 
and in the child component create a 
```elm
module ComplicatedComponent exposing (Msg(..), IMsg) -- expose the constructors for the "public" events, hide the "internal" IMsg events forcing callers to pass them into a locally defined `update`

type Msg
    = Clicked -- public, exposed constructor we'll force callers to handle because exhaustivity (woo Elm)
    | IMsg IMsg -- private, non exposed constructors we have to handle internally.
    
type IMsg -- mostly splitballing some random constructors.
    = Closed
    | Hovered
    | Scrolled
    
then you're cumbersome "Translator" / "Builder" becomes

view : Model -> Html Msg -- notice the capital Msg constructor here.
update : IMsg -> Model -> (Model, Cmd Msg) -- notice we're NOT handling the exposed Clicked msg...
```
This does a few things, yes it forces a `Cmd.map Wrapper` in a couple places, BUT the callers no longer have to define their own Msg constructors to wire into `Stuff -> msg` arguments to satisfy the interface. AND the signatures for the `view` / `update` for the "component" become much simpler.

## Now. to answer your question more specifically.
As far as Shared state is concerned, which the "Store" pattern is inherently shared state. Yes that'll need to be a top-level module. BUT! to fix the "wiring" problem for components you want to be able to trigger these Shared _Effect_'s from. 

You want

# Effect
Except not [this Effect](http://sporto.github.io/elm-patterns/architecture/effects.html) altho in principle it's sorta similar...
[THIS EFFECT](https://github.com/erlandsona/elm-ui-form/blob/main/src/Effect.elm)
Which get's converted back into a top-level command via Main.elm
<img width="687" alt="Screen Shot 2022-06-25 at 8 25 39 AM" src="https://user-images.githubusercontent.com/10421138/175775475-a430b8af-17d7-48d4-88b2-a571378623a6.png">


What this does is you change all you're Page updates from `Msg -> Model -> (Model, Cmd Msg)` to `Msg -> Model -> (Model, Effect Msg)` which enables Pages to emit `Shared.Msg`'s side by side with their own Msg's...

---

Now before I mentioned there's a few ways to approach the "component" thing... some "components" actually do deserve internal state but we don't really care to build all the wiring for them... Good example is a Tooltip. Sure we can turn the state for a Tooltip into a Sum type to "ensure only one is open at a time" but in my experience that's more boilerplate than it's worth and ends up not scaling to more than a few tooltips for a given page before you have to copy/paste the pattern rinse and repeat and sometimes I just want to say `Html.div [ Attr.tooltip "Some Context" ] []` or whatever, right?

In this case, you can actually build it like that but the Attribute approach won't have access to say the `window` dimensions for ensuring it doesn't go off the edge of the screen.

## webcomponents

This is when you should probably start digging into leveraging a webcomponent.
webcomponents dramatically simplify the Elm-side api because it becomes a dumb view component `Html.node "tooltip" [] []`. The flip side to this is now you're writing all your UI logic in JS so they're super finicky. 

But instantiating them anywhere and everywhere you need them on the Elm side becomes a trivial view concern.

**Comment by eratio08** [comment]
Thanks for your extensive and detailed answer! 
If I understand this correctly my nested-TEA-message-issue is solvable using the Shared.Msg Effects. I will apply this and see where it takes me!
